pkg/state/state.go

```diff
@@ -354,11 +354,26 @@ func New(ctx context.Context, pgURL, stateSchema string) (*State, error) {
 }
 
 func (s *State) Init(ctx context.Context) error {
-	// ensure pgroll internal tables exist
-	// TODO: eventually use migrations for this instead of hardcoding
-	_, err := s.pgConn.ExecContext(ctx, fmt.Sprintf(sqlInit, pq.QuoteIdentifier(s.schema)))
+	tx, err := s.pgConn.Begin()
```

Comment:
do we really need a transaction? I'm wondering if this could cause locking for other table modifications, for instance if some triggers are using our functions

Reply:
I was unable to get the same liveness guarantees using a session level lock (`pg_advisory_lock`) as I was using the transaction level locks. Using the diff below and the test program from [this gist](https://gist.github.com/andrew-farries/883a9f314834067ca3301548844c7d1b) I was able to observe instances where all goroutines would become blocked or where Postgres would report deadlocks.

```diff
diff --git a/pkg/state/state.go b/pkg/state/state.go
index 79f3597..9eb1248 100644
--- a/pkg/state/state.go
+++ b/pkg/state/state.go
@@ -354,28 +354,23 @@ func New(ctx context.Context, pgURL, stateSchema string) (*State, error) {
 }
 
 func (s *State) Init(ctx context.Context) error {
-	tx, err := s.pgConn.Begin()
-	if err != nil {
-		return err
-	}
-	defer tx.Rollback()
-
 	// Try to obtain an advisory lock.
 	// The key is an arbitrary number, used to distinguish the lock from other locks.
 	// The lock is automatically released when the transaction is committed or rolled back.
 	const key int64 = 0x2c03057fb9525b
-	_, err = tx.ExecContext(ctx, "SELECT pg_advisory_xact_lock($1)", key)
+	_, err := s.pgConn.ExecContext(ctx, "SELECT pg_advisory_lock($1)", key)
 	if err != nil {
 		return err
 	}
 
+	defer func() {
+		s.pgConn.ExecContext(ctx, "SELECT pg_advisory_unlock($1)", key)
+	}()
+
 	// Perform pgroll state initialization
-	_, err = tx.ExecContext(ctx, fmt.Sprintf(sqlInit, pq.QuoteIdentifier(s.schema)))
-	if err != nil {
-		return err
-	}
+	_, err = s.pgConn.ExecContext(ctx, fmt.Sprintf(sqlInit, pq.QuoteIdentifier(s.schema)))
 
-	return tx.Commit()
+	return err
 }
 
 func (s *State) Close() error {

```